### PR TITLE
MIGRATE multikey and filter implementation

### LINKED ISSUE #1423: Jedis.migrate() don't support "replace" param
### API probleam
As in `redis-trib.rb`, when you got a `BUSYKEY` exception, you need to replace the node using `migrate`.
image

In Jedis `migrate` method don't have the `replace` param.
```
public String migrate(final String host, final int port, final String key, final int destinationDb, final int timeout)
```

#### Jedis version:
jedis-2.8

### PR DESCRIPTION
**Pull request by sazzad16**
Closes #1423

### CODE REVIEW

File: src/main/java/redis/clients/jedis/BinaryClient.java [hunk 1]
```diff
@@ -1030,10 +1032,28 @@ public void time() {
     sendCommand(TIME);
   }
 
-  public void migrate(final byte[] host, final int port, final byte[] key, final int destinationDb,
-      final int timeout) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryClient.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/BinaryJedis.java [hunk 2]
```diff
@@ -3501,10 +3502,19 @@ public String clientSetname(final byte[] name) {
     return client.getMultiBulkReply();
   }
 
-  public String migrate(final byte[] host, final int port, final byte[] key,
-      final int destinationDb, final int timeout) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryJedis.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/Client.java [hunk 3]
```diff
@@ -881,9 +882,16 @@ public void clientSetname(final String name) {
     clientSetname(SafeEncoder.encode(name));
   }
 
-  public void migrate(final String host, final int port, final String key, final int destinationDb,
-      final int timeout) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Client.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 4]
```diff
@@ -3159,10 +3160,19 @@ public String clientSetname(final String name) {
     return client.getStatusCodeReply();
   }
 
+  @Override
   public String migrate(final String host, final int port, final String key,
-      final int destinationDb, final int timeout) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/PipelineBase.java [hunk 5]
```diff
@@ -1399,13 +1399,17 @@
     return getResponse(BuilderFactory.BYTE_ARRAY);
   }
 
-  public Response<String> migrate(final String host, final int port, final String key, final int destinationDb, final int timeout) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/PipelineBase.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/PipelineBase.java [hunk 6]
```diff
@@ -1399,13 +1399,17 @@
...
-    getClient(key).migrate(host, port, key, destinationDb, timeout);
+  @Override
+  public Response<String> migrate(final String host, final int port, final String key,
+      final int destinationDB, final long timeout) {
+    getClient(key).migrate(host, port, key, destinationDB, timeout);
     return getResponse(BuilderFactory.STRING);
   }
 
-  public Response<String> migrate(final byte[] host, final int port, final byte[] key, final int destinationDb, final int timeout) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/PipelineBase.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/PipelineBase.java [hunk 7]
```diff
@@ -1399,12 +1399,16 @@
...
 
-  public Response<String> migrate(final String host, final int port, final String key, final int destinationDb, final int timeout) {
+  @Override
+  public Response<String> migrate(final String host, final int port,
+      final String key, final int destinationDb, final int timeout) {
     getClient(key).migrate(host, port, key, destinationDb, timeout);
     return getResponse(BuilderFactory.STRING);
   }
 
-  public Response<String> migrate(final byte[] host, final int port, final byte[] key, final int destinationDb, final int timeout) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/PipelineBase.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/BinaryJedis.java [hunk 8]
```diff
@@ -3501,13 +3502,22 @@ public String clientSetname(final byte[] name) {
     return client.getMultiBulkReply();
   }
 
-  public String migrate(final byte[] host, final int port, final byte[] key,
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryJedis.java]: Requires deprecation.

File: src/main/java/redis/clients/jedis/BinaryClient.java [hunk 9]
```diff
@@ -1030,10 +1031,28 @@ public void time() {
     sendCommand(TIME);
   }
 
-  public void migrate(final byte[] host, final int port, final byte[] key, final int destinationDb,
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BinaryClient.java]: Requires deprecation.

**Comment by sazzad16** [comment]
@HeartSaVioR @marcosnils ?

**Comment by marcosnils** [comment]
@sazzad16 can you please rebase to check that tests pass?. Thx.

**Comment by sazzad16** [comment]
@marcosnils rebased. tests pass.

**Comment by sazzad16** [comment]
@marcosnils It's 3.0 because package `redis.clients.jedis.params` just introduced in 3.0 and not available for earlier branches.

Moreover, host is always received as String except for `migrate` where it is byte[]. To be strict to one pattern, I've changed all migrate to receive host as String. Deprecation is required because of this.

**Comment by sazzad16** [comment]
Merged into master. Downmerged with deprecation into 2.10.

**Pull request merged**